gui/src/vmm/cpu/x86_64.rs

```diff
@@ -31,11 +36,60 @@ impl<H: Hypervisor, S: Screen> Breakpoints for CpuManager<H, S> {
 }
 
 impl<H: Hypervisor, S: Screen> SwBreakpoint for CpuManager<H, S> {
-    fn add_sw_breakpoint(&mut self, addr: u64, kind: usize) -> TargetResult<bool, Self> {
-        todo!()
+    fn add_sw_breakpoint(&mut self, addr: u64, _kind: usize) -> TargetResult<bool, Self> {
+        let Entry::Vacant(entry) = self.sw_breakpoints.entry(addr) else {
+            return Ok(false);
+        };
+
+        let cpu = self.cpus.first_mut().unwrap();
+
+        let translated_addr = cpu
+            .debug_mut()
+            .unwrap()
+            .translate_address(addr.try_into().unwrap())
+            .ok_or(TargetError::Fatal(GdbError::MainCpuExited))?;
+
+        // Get data.
+        let mut src = self
+            .hv
+            .ram()
+            .lock(translated_addr, BREAKPOINT_SIZE)
+            .ok_or(TargetError::Errno(Self::GDB_EFAULT))?;
+
+        let code_slice = src.as_mut_ptr().cast();
+
+        let mut code_bytes = std::mem::replace(&mut unsafe { *code_slice }, BREAKPOINT_BYTES);
```

Comment:
I'm feeling unsafe here since `BREAKPOINT_SIZE` and `BREAKPOINT_BYTES` is a separated variable. Better to remove the whole `BREAKPOINT_BYTES` and hard-coded length of data to lock to `1`. Then use `std::mem::replace` on a single `u8` instead of an array. Then we can use `entry.insert([replaced])`, which will be automatically get type-check if someone change `BREAKPOINT_SIZE` to something else.